src/message.ts

```diff
@@ -0,0 +1,79 @@
+import { Promise } from 'es6-promise';
+import { TALKER_TYPE } from './strings';
+import Talker from './index';
+
+abstract class Message {
+  protected readonly type: string = TALKER_TYPE;
+
+  constructor(
+    /*
+     * @property talker - A {@link Talker} instance that will be used to send responses
+     */
+    protected readonly talker: Talker,
+    /*
+     * @property namespace - A namespace to with which to categorize messages
+     */
+    public readonly namespace: string,
+    public readonly data: Stringifyable,
+    public readonly responseToId: number | null = null
+  ) {}
+}
+
+export interface JSONifiedMessage {
+  readonly namespace?: string;
+  readonly data?: Stringifyable;
+  readonly id?: number;
+  readonly responseToId?: number;
+  readonly type: string;
+  readonly handshake?: boolean;
+  readonly handshakeConfirmation?: boolean;
+}
+
+export interface Stringifyable {
+  [index: string]: string | number | Stringifyable | Stringifyable[] | boolean | null | undefined;
+}
+
+// Consuming applications will almost never interact with this class.
+export class OutgoingMessage extends Message {
+  public readonly id: number = this.talker.nextId();
+
+  /**
+   * @param talker
+   * @param namespace
+   * @param data
+   * @param responseToId - If this is a response to a previous message, its ID.
+   */
+  constructor(
+    protected readonly talker: Talker,
+    public readonly namespace: string,
+    public readonly data: Stringifyable,
+    public readonly responseToId: number | null = null
+  ) {
+    super(talker, namespace, data, responseToId);
+  }
+
+  toJSON(): JSONifiedMessage {
```

Comment:
Unless `toJSON` has some special meaning.

```suggestion
  toJSONable(): JSONableMessage {
```

or 

```suggestion
  toSerializable(): SerializableMessage {
```

Reply:
I'd like to avoid bumping the major version number with this PR, so I don't want to change the public interface.

Reply:
Also Web APIs usually have a `toJSON` method that returns an object, not a string. https://developer.mozilla.org/en-US/docs/Web/API/Performance/toJSON / https://developer.mozilla.org/en-US/docs/Web/API/PushSubscription/toJSON / https://developer.mozilla.org/en-US/docs/Web/API/PaymentAddress/toJSON etc.

Reply:
I didn't know that! Interesting terminology.